package-res/resources/web/pentaho/type/property.js

```diff
@@ -0,0 +1,1331 @@
+           *
+           * When set and the property already has [descendant]{@link pentaho.type.Type#hasDescendants} properties,
+           * an error is thrown.
+           *
+           * The default (root) `isApplicable` attribute value is `true`.
+           *
+           * @name isApplicable
+           * @memberOf pentaho.type.Property.Type#
+           * @type undefined | boolean | pentaho.type.PropertyDynamicAttribute.<boolean>
+           *
+           * @see pentaho.type.Property.Type#isRequired
+           * @see pentaho.type.Complex#isApplicable
+           * @see pentaho.type.spec.IPropertyTypeProto#isApplicable
+           */
+          isApplicable: {
+            value: true,
+            cast: Boolean,
+            combine: function(baseEval, localEval) {
+              return function() {
+                // localEval is skipped if base is false.
+                return baseEval.call(this) && localEval.call(this);
+              };
+            }
+          },
+
+          /**
+           * Evaluates the value of the `isReadOnly` attribute of a property of this type
+           * on a given owner complex value.
+           *
+           * @name isReadOnlyEval
+           * @memberOf pentaho.type.Property.Type#
+           * @param {pentaho.type.Complex} owner The complex value that owns a property of this type.
+           * @return {boolean} The evaluated value of the `isReadOnly` attribute.
+           *
+           * @ignore
+           */
+
+          /**
+           * Gets or sets a value, or function, that indicates if properties of this type
+           * _cannot_ be changed by a user, in a user interface.
+           *
+           * A property should be set read-only whenever its value is implied/imposed somehow,
+           * and thus cannot not be changed, directly, by the user, in a user interface.
+           *
+           * ### This attribute is *Dynamic*
+           *
+           * When a _dynamic_ attribute is set to a function,
+           * it can evaluate to a different value for each complex instance.
+           *
+           * When a _dynamic_ attribute is set to a value other than a function or a {@link Nully} value,
+           * its value is the same for every complex instance.
+           *
+           * ### This attribute is *Monotonic*
+           *
+           * The value of a _monotonic_ attribute can change, but only in some, predetermined _monotonic_ direction.
+           *
+           * In this case, a _property type_ marked as _not read-only_ can later be marked as _read-only_.
+           * However, a _property type_ marked as _read-only_ can no longer go back to being _not read-only_.
+           *
+           * Because this attribute is also _dynamic_,
+           * the actual `isReadOnly` values are only known
+           * when evaluated for specific complex instances.
+           * As such, ensuring monotonic changes is deferred until evaluation.
```

Comment:
Suggest:

This behavior ensures that monotonic changes are deferred until evaluation.